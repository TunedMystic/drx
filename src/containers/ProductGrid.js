import React, { Component } from 'react';
import { connect } from 'react-redux';
import { isEmpty } from 'lodash';

import { fetchProducts, removeProduct } from '../actions/products';

class ProductGrid extends Component {
    componentDidMount() {
        this.initialFetch();
    }

    initialFetch = () => {
        if (isEmpty(this.props.products.productsById)) {
            this.props.fetchProducts();
        }
    }

    render() {
        const { productsById, productIds } = this.props.products;
        return (
            <div>
                <h3>Product Grid</h3>
                <ul>                
                    {
                        productIds.map(id => (
                            <li key={id}>{productsById[id].name}</li>
                        ))
                    }
                </ul>
            </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => ({
    products: state.products,
    ...ownProps
});

const mapActionsToProps = {
    fetchProducts,
    removeProduct
};

export default connect(mapStateToProps, mapActionsToProps)(ProductGrid);
