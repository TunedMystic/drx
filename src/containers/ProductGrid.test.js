import axios from 'axios';
import { cleanup, waitForElement } from 'react-testing-library';

import renderWithRedux from '../utils/render-with-redux';
import ProductGrid from './ProductGrid';

jest.mock('axios');

afterEach(cleanup);

describe('Container', () => {
    let productsResponse = [
        {
            id: 'mvx7',
            name: 'Pizza',
            price: 5.99
        },
        {
            id: 'hyt3',
            name: 'Cake',
            price: 12.67
        },
        {
            id: 'k85x',
            name: 'Salad',
            price: 4.78
        },
    ];

    productsResponse = {data: {json: productsResponse}};

    it('should render correctly', async () => {
        axios.post.mockReturnValue(new Promise(resolve => resolve(productsResponse)));
        const { container } = renderWithRedux(<ProductGrid />);
        expect(container).toBeInTheDocument();
    });

    it('should match snapshot', async () => {
        axios.post.mockReturnValue(new Promise(resolve => resolve(productsResponse)));
        const { container } = renderWithRedux(<ProductGrid />);
        expect(container).toMatchSnapshot();
    });

    it('should render <Product /> components', async () => {
        axios.post.mockReturnValue(new Promise(resolve => resolve(productsResponse)));
        const { queryByText, getAllByTestId } = renderWithRedux(<ProductGrid />);

        // Wait for the data to resolve.
        // Isn't the data resolved immediately? Is this really helpful?
        await waitForElement(() => getAllByTestId('product'));

        // Check that 3 products have rendered.
        expect(getAllByTestId('product')).toHaveLength(3);

        expect(queryByText('Pizza')).toBeInTheDocument();
        expect(queryByText('Cake')).toBeInTheDocument();
        expect(queryByText('Salad')).toBeInTheDocument();
    });
});
