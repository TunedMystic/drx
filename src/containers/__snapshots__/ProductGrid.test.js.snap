// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container should match snapshot 1`] = `
<div>
  <div
    class="jsx-3403093102 product-grid"
    data-testid="product-grid"
  />
</div>
`;
